Pattern matching syntax and semantics

zygoloid opened the pull request:
This paper proposes concrete syntax and semantic choices for Carbon patterns.

zygoloid committed:
Proposal: pattern matching syntax and semantics.

zygoloid committed:
Pre-commit.

zygoloid committed:
Allow `{.a: T, ...}` as a pattern even though it contains no bindings.

zygoloid committed:
Avoid mismatched braces in examples.

zygoloid committed:
Handle grouping parentheses.

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+A name binding is a pattern.
+
+-   _binding-pattern_ ::= `unused`? _identifier_ `:` _expression_
+-   _binding-pattern_ ::= `unused`? _identifier_ `:` _proper-pattern_
```
I feel like this second form is controversial. Is it required?

zygoloid replied:
We could probably replace it with _identifier_ `:` `auto` (and remove `auto` patterns) without affecting many of our current examples. I think this syntax could be valuable in some cases, for example for type-dependent `match`es in templates:

```
fn F[template T:! Type](x: T) -> auto {
  match (x) {
    case (a: (U:! Type), b: (V:! AddableWith(U))) => {
      return F(a + b);
    }
    // ...
  }
}
```

... but if we'd rather not allow this (and maybe permit `case forall` if we find things like the above motivating), I don't think it creates any major problems to remove it. (explorer has only a few testcases using this.)

josh11b replied:
My main arguments against it are readability and consistency with the decision we made for `impl` declarations. How does the alternative read?
```
fn F[template T:! Type](x: T) -> auto {
  match (x) {
    case forall [U:! Type, V:! AddableWith(U)] (a: U, b: V) => {
      return F(a + b);
    }
    // ...
  }
}
```

@mconst

mconst replied:
I think I don't have a strong preference here.  I do find `case (a: (U:! Type), b: (V:! AddableWith(U)))` hard to read, but I'm not sure `case forall [U:! Type, V:! AddableWith(U)] (a: U, b: V)` is significantly better, especially since it requires the unusual `forall` keyword.  So either way's fine with me.

zygoloid replied:
On discord, @chandlerc seemed pretty strongly in the camp of not permitting patterns in this context, so I'm going with that direction for now. I think the consequence is that we are effectively separating pattern matching from type deduction, and treating them as two different things, and I think that's reasonable. (That's what explorer does, for what it's worth.)

Done.

micttyl commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+
+### Expressions versus proper patterns
+
+A pattern that contains within it any pattern-specific syntax, such as a
```
Can you rephrase this?

zygoloid replied:
Done. Better?

micttyl replied:
It is clear.

micttyl commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+
+### `auto`
+
+The pattern `auto` is shorthand for `_:! Type`.
```
Treat them equivalent?
```
var v: auto = SomeComplicatedExpression(n);
var v: (_:! Type) = SomeComplicatedExpression(n);
```

zygoloid replied:
Added as an example.

micttyl replied:
It looks good.

micttyl commented on proposals/p2188.md:
````diff
@@ -0,0 +1,456 @@
+  }
+}
+```
+
````
No dynamic type yet?

zygoloid replied:
Added as future work.

micttyl replied:
it looks good.

micttyl commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+A `var` pattern matches when its nested pattern matches. The type of the storage
+is the resolved type of the nested _pattern_. Any bindings within the nested
+pattern refer to portions of the corresponding storage rather than to the
+scrutinee, somewhat as if the mutable storage is first initialized from the
```
Can you rephrase this?

zygoloid replied:
Better?

micttyl replied:
lgtm

micttyl reviewed (commented):
I need to read more to understand. This is because I haven't caught up the current state of Carbon.

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+
+### `auto`
+
+The pattern `auto` is shorthand for `_:! Type`.
```
This isn't what I would expect in a function parameter, where I would expect it to be a template.

zygoloid replied:
Typo: I meant for it to be a template parameter in general. Fixed.

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+A tuple of patterns can be used as a pattern.
+
+-   _tuple-pattern_ ::= `(` [_expression_ `,`]\* _proper-pattern_ [`,`
+    _pattern_]\* `)`
```
Allow optional trailing `,`, which is required to make a 1-tuple?

zygoloid replied:
Trailing `,` for one-tuples is an [open question](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/tuples.md#single-value-tuples). Filed #2191, but I've changed this proposal to allow trailing commas on the assumption that that issue will go the way I think it will :)

zygoloid replied:
#2191 decided: trailing `,` is allowed.

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+A choice pattern is used to match one alternative of a choice type.
+
+_proper-pattern_ ::= _callee-expression_ _tuple-pattern_ _proper-pattern_ ::=
+_designator_ _tuple-pattern_?
```
```suggestion
-   _proper-pattern_ ::= _callee-expression_ _tuple-pattern_
-   _proper-pattern_ ::= _designator_ _tuple-pattern_?
```

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+other struct syntaxes:
+
+-   `{.field: type}` seems like a contender, but doesn't work because that is
+    already recognized as a tuple type literal. Also, the use of normal binding
```
```suggestion
    already recognized as a struct type literal. Also, the use of normal binding
```

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+    `{.field = field: type}`, but looks a little surprising. This is probably
+    the best choice if concerns are found with `{field: type}` syntax.
+
+## Future work
```
Dynamic type tests?

zygoloid replied:
Added.

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,456 @@
+Note that a pattern of the form `Optional(T).None` is an expression pattern and
+is compared using `==`.
+
+### Templates
```
We're going to need to have a conversation about whether this should also be allowed for values with checked/symbolic generic types.

zygoloid replied:
Acknowledged. For now I want to reject, given that dispatch based on the actual type seems very template-y.

zygoloid committed:
Add bullets to prevent grammar productions from wrapping together

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

zygoloid committed:
Fix tuple / struct typo.

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

zygoloid committed:
Allow trailing comma for one-tuple patterns.

zygoloid committed:
Allow reuse of comparison results. Improve wording. Add future work to
support dynamic type matching.

micttyl reviewed (approved):
It is clear.

geoffromer commented on proposals/p2188.md:
```diff
@@ -0,0 +1,517 @@
+    scrutinee.
+-   Additional syntax to make matching against structs more convenient.
+
+## Details
```
This proposal doesn't cover deduced parameters, even as future work, but I've been thinking of them as an integral part of pattern matching. Do you consider them to be separate?

zygoloid replied:
I think the discussion on `name: pattern` seems to have converged on type deduction syntax being separate from pattern matching syntax. The semantic models behind them might be the same, but we don't seem to like using the same syntax for both.

geoffromer replied:
My point is more about semantics than syntax, but let's move this discussion to my comment near the bottom.

geoffromer commented on proposals/p2188.md:
```diff
@@ -0,0 +1,517 @@
+    `{.field = field: type}`, but looks a little surprising. This is probably
+    the best choice if concerns are found with `{field: type}` syntax.
+
+## Future work
```
Deduced parameters?

zygoloid replied:
Added.

geoffromer commented on proposals/p2188.md:
````diff
@@ -0,0 +1,517 @@
+```
+
+Any `==` operations performed by a pattern match occur in lexical order, but for
+repeated matches against the same `_pattern_`, later comparisons may be skipped
````
```suggestion
repeated matches against the same _pattern_, later comparisons may be skipped
```

geoffromer commented on proposals/p2188.md:
````diff
@@ -0,0 +1,517 @@
+}
+```
+
+The `unused` keyword indicates that the binding is intended to not be used.
````
It might be a good idea to explicitly say that this is intended to restate #2022's decision, not supersede it, just in case we miss some subtle disagreement between them.

zygoloid replied:
Done.

geoffromer commented on proposals/p2188.md:
```diff
@@ -0,0 +1,517 @@
+field names as the pattern. The pattern is matched left-to-right, meaning that
+matching is performed in the field order specified in the pattern, not in the
+field order of the scrutinee. This is consistent with the behavior of matching
+against a tuple-valued expression, where the left operand determines the order
```
```suggestion
against a struct-valued expression, where the left operand determines the order
```

geoffromer commented on proposals/p2188.md:
```diff
@@ -0,0 +1,517 @@
+field names as the pattern. The pattern is matched left-to-right, meaning that
+matching is performed in the field order specified in the pattern, not in the
+field order of the scrutinee. This is consistent with the behavior of matching
+against a tuple-valued expression, where the left operand determines the order
```
"_pattern_ operand" might be clearer; as it is, I couldn't remember which operand was on the left in the expression case.

zygoloid replied:
Done.

geoffromer commented on proposals/p2188.md:
````diff
@@ -0,0 +1,517 @@
+}
+```
+
+If some fields should be ignored when matching, a trailing `, ...` can be added
````
I worry that this could be mistaken for a variadic syntax, or perhaps even be ambiguous with one.

As you may recall, the current (very preliminary) plan is to express pack expansion using special operator tokens that start with `...`, such as `...,` and `...{`. I don't anticipate wanting a `...}` operator, so there's no outright ambiguity, but it could easily cause confusion for human readers. Furthermore, we were previously considering using a _trailing_ `...` for those operators, in which case `,...` would be ambiguous between this syntax and the comma pack expansion. I currently think leading `...` is a better choice, but switching back does seem within the realm of possibility.

How about something like `._` as an alternative syntax? It's not perfect, because it looks like it should match exactly one field, but it's the best I've come up with so far.

OTOH I'm not very satisfied with the variadic syntax as a whole, so I'm OK with keeping this struct pattern syntax for now, in the hope that we can come up with a better variadic syntax. I just want to make sure this issue is considered.

zygoloid replied:
I'm finding that I like `, _` more than `, ...` or `, ._`, even if we don't end up using `...` for variadics, so I've switched to that for now.

mconst replied:
Could we potentially make this match the variadic syntax better, by using something like `..., _`?  That way it looks like a pack expansion of `_`, which seems like a pretty good analogy for what it's doing.

zygoloid replied:
Potentially. That seems like something to consider as part of the variadics proposal... though that syntax seems a bit cumbersome to me, and it might suggest that you could use just `, _` to ignore exactly one field, which I don't think is an interesting facility to provide.

geoffromer commented on proposals/p2188.md:
```diff
@@ -0,0 +1,517 @@
+
+This is valid even if all fields are actually named in the pattern.
+
+### Choice patterns
```
I'd be inclined to call these "alternative patterns", since they have the form and meaning of a single alternative, not a whole `choice` type.

zygoloid replied:
Done.

geoffromer commented on proposals/p2188.md:
```diff
@@ -0,0 +1,517 @@
+
+### User-defined pattern matching
+
+We could provide some mechanism for allowing a user-defined type to specify how
+it can be matched by patterns.
```
This is already more or less the plan of record.

```suggestion
We plan to provide a mechanism for allowing a user-defined type to specify how
it can be matched by patterns. See [p0157.md](p0157) for details.
```

micttyl commented on proposals/p2188.md:
````diff
@@ -0,0 +1,517 @@
+}
+```
+
+The `unused` keyword indicates that the binding is intended to not be used.
````
Can you show the use case of  `unused`

zygoloid replied:
There is now an example of this a little later in the proposal, plus a cross-reference to #2022 which introduces this keyword.

zygoloid committed:
Address review comments and feedback. Remove `identifier: pattern`.

zygoloid committed:
Add ref to #157.

Co-authored-by: Geoff Romer <gromer@google.com>

zygoloid committed:
Fix xref.

micttyl reviewed (approved):
Thank you. Understood. Confusing though.

zygoloid committed:
Add more xrefs to #2022 and to #1084.

geoffromer commented on proposals/p2188.md:
```diff
@@ -515,3 +655,11 @@ consistent with the behavior of `==` in the case where a struct is implicitly
+### Type deduction
+
+This proposal does not cover type deduction, instead considering it to be a
+separate topic from pattern matching syntax, even though the semantic behavior
+of the two may be quite similar or identical.
```
This is a proposal for pattern matching syntax _and_ semantics, though. Even at the syntactic level, they're closely connected: a deduced parameter list always comes immediately before a pattern, and always consists of a list of binding patterns. At the semantic level, it's not just that the behavior of the two is "similar", it's that they are directly coupled: as far as I can tell, type deduction is always driven by a pattern matching operation, and I think it can feed back into that pattern matching operation.

To be clear, I'm not saying this proposal needs to have a design for type deduction; I just don't think type deduction is as cleanly separable from pattern matching as this makes it sound.

zygoloid replied:
I don't think they're *necessarily* coupled. I think we are confident that we're going to do deduction when performing function calls to generic functions and when selecting an `impl`. I think we are confident that we're going to do pattern matching in a `match` statement and in a `let` or `var` binding. I don't think it's guaranteed that we'll do pattern matching in a function call -- it may be reasonable to restrict a function to have only a list of bindings as its parameters, which only needs pattern matching in a degenerate sense, and I think we're not going to do pattern matching in `impl` selection. So I think there might actually be no situations where we perform both type deduction and non-degenerate pattern matching.

I also think it's possible that we'll model the matching process for the two cases quite differently: for pattern matching, we have a top-down process where the pattern syntax determines what happens at each level (such as evaluating an expression pattern and performing an `==` comparison, or converting the scrutinee), whereas for type deduction I think it's likely we'll instead perform a symbolic evaluation of the type expression (the "pattern") and perform a value comparison between (symbolic) type values to find a set of substitutions. The difference is fairly subtle, but I think deduction ends up having stricter matching semantics than a pattern match -- identical after substitution versus a looser notion of "match".

geoffromer replied:
> So I think there might actually be no situations where we perform both type deduction and non-degenerate pattern matching.

It seems like variadic functions will be such a situation. Maybe we can say that semantically, variadics relate only to deduction, and not pattern matching (although I'm not sure of that), but syntactically they're going to appear in the parentheses as well as the square brackets.

geoffromer commented on proposals/p2188.md:
```diff
@@ -168,17 +161,21 @@ fn G() -> i32 {
 
-The `unused` keyword indicates that the binding is intended to not be used.
+As specified in
+[#2022](https://github.com/carbon-language/carbon-lang/pull/2022), the `unused`
```
```suggestion
As specified in [p2022](p2022.md), the `unused`
```

geoffromer commented on proposals/p2188.md:
```diff
@@ -472,6 +525,93 @@ other struct syntaxes:
+for now:
+
+-   In pattern matching, bindings and wildcards are introduced by nested `:` /
+    `:!` patterns.
```
```suggestion
-   In pattern matching, bindings and wildcards are introduced by nested `:` /
    `:!` patterns, and the right-hand side of a binding pattern is never a proper
    pattern.
```

zygoloid committed:
Update proposals/p2188.md

Co-authored-by: Geoff Romer <gromer@google.com>

zygoloid committed:
Update links now #2022 has landed.

zygoloid committed:
Reformat

zygoloid committed:
Merge branch 'trunk' into proposal-pattern-match

josh11b commented on proposals/p2188.md:
```diff
@@ -0,0 +1,664 @@
+### Templates
+
+If the type of the scrutinee of a pattern is either a template parameter or is
+an associated type that depends on a template parameter, any checking of the
```
What about other expressions dependent on a template parameter like `template T:! type, v: Vector(T)`?

zygoloid replied:
My intent was that we would be able to reject a tuple or struct pattern that tries to match `v` when checking the template definition. Hm, maybe that doesn't play well with the phases model for templates: if we don't symbolically evaluate `Vector(T)` because it involves a template parameter, then I guess we can't really ask whether it's a tuple or struct type.

josh11b replied:
Maybe the definition of dependent from #2200 could be used here?

zygoloid replied:
I'm not sure that helps. For example, if we have `a: Vector(T)` and `b: Identity(T)`, I would expect both to have the same properties -- both type expressions are known to be of type `Type`, and for both of them the actual type value depends on a template parameter -- but for `a` we might know something about how to do pattern matching whereas for `b` we do not.

Also, given the extensibility support in https://github.com/carbon-language/carbon-lang/pull/2187, I suppose it might be the case that even for `a` we don't actually know how to do matching until we know `T` -- the `impl as Match` that's used might depend on `T`.

josh11b commented on proposals/p2188.md:
````diff
@@ -0,0 +1,664 @@
+```
+fn NeverWorks[template T:! Type](triple: (T, T, T)) {
+  match (triple) {
+    // ❌ Error: type mismatch matching struct against (T, T, T)
````
Would this error be diagnosed at definition time or instantiation?

zygoloid replied:
My intent was to diagnose at template definition time. But per the previous comment I guess maybe that doesn't fit the model well. Removed.

josh11b replied:
I think we should be able to say that these differences are diagnosed at definition time.

zygoloid replied:
OK, I've added this back. As noted in the previous comment, I don't think the definition of dependent from #2200 is sufficient here, but I think it's sufficient in this design doc to describe the intent -- if it's clear in the template definition that the pattern can never match, then that's an error -- and omit the case analysis with the details.

zygoloid committed:
Don't try to reason symbolically about types in templates.

camio commented on proposals/p2188.md:
```diff
@@ -0,0 +1,655 @@
+expressions, patterns add:
+
+-   Bindings, of the form `name: type`, which give a name for the scrutinee.
+-   `var` _pattern_, which requests mutable storage be provided for the
```
It is unclear what is meant by mutable storages "provide for" the scrutinee.

zygoloid replied:
Reworded.

camio commented on proposals/p2188.md:
````diff
@@ -0,0 +1,655 @@
+}
+```
+
+As specified in [#2022](p2022.md#the-behavior-of-unused-name-bindings), the
````
This link doesn't seem to work. Maybe use an absolute path instead? Same goes for the other links.

zygoloid replied:
I _think_ this is a consequence of viewing this from a github PR and it would work when checked in, but fixed anyway, just in case.

camio commented on proposals/p2188.md:
````diff
@@ -0,0 +1,655 @@
+
+```
+fn F(n: i32) {
+  var v: auto = SomeComplicatedExpression(n);
````
Have you considered introducing a shorthand where `auto` can be omitted?

```
var v:= SomeComplicatedExpression(n);
```

zygoloid replied:
#851 considered some such shorthands, but not that one. Added to "Alternatives considered". I think that syntax is probably my favourite so far if we find we want a shorthand for this purpose, but I don't think it's urgent to add a shorter syntax now. We can evolve in this direction in the future if we find a sustained interest in it.

camio commented on proposals/p2188.md:
```diff
@@ -0,0 +1,655 @@
+-   _proper-pattern_ ::= _binding-pattern_
+
+The type of the _identifier_ is specified by the _expression_. The scrutinee is
+implicitly converted to that type.
```
"implicitly converted to that type" implies that there is a copy happening here. Is that what the intent is? It would be good to have examples that take the address of a binding and compare it to the address of the scrutinee. For efficiency I think it is important to guarantee that there are no copies when no conversion is necessary.

zygoloid replied:
I agree that we should have stronger guarantees here. I've added some words and notably some examples.

We don't allow `let` bindings to have their addresses taken, and I don't think we intend to guarantee they have a stable address (or at least, we've not agreed that they do), so I don't think I can easily provide the example you asked for, but I can demonstrate when destructors are run at least.

camio commented on proposals/p2188.md:
````diff
@@ -0,0 +1,655 @@
+}
+```
+
+If some fields should be ignored when matching, a trailing `, _` can be added to
````
This is a nice rule.

camio commented on proposals/p2188.md:
```diff
@@ -0,0 +1,655 @@
+potentially reducing the number of rules that Carbon developers need to learn.
+
+We find that attempting to unify pattern-matching and type deduction in this way
+harms readability. Having distinct syntax for type-level matching and
```
I don't know that I disagree that the unification harms readability, but your examples are contrived so it's really hard to imagine how this would impact typical code.

zygoloid replied:
Weakened the claim and added an example.

camio commented on proposals/p2188.md:
```diff
@@ -0,0 +1,655 @@
+### Guards
+
+We could provide guards for patterns, allowing a pattern to match only if some
+predicate involving the bindings holds:
```
A pattern matching syntax without guards is very difficult to use when a guard is needed. It seems to me like this should be part of the base set of operations and its low complexity doesn't seem to warrant an initial exclusion.

zygoloid replied:
OK, I'm sold, added.

camio commented on proposals/p2188.md:
```diff
@@ -0,0 +1,655 @@
+
+We could provide a way to match polymorphic class objects based on their dynamic
+type. This might be the default when matching a polymorphic class, or might
+require opt-in.
```
I think it is important to explore this at least a bit. It isn't clear that the current design can easily support an extension like this.

zygoloid replied:
Added a little exploration with the goal of showing that this can be done within this syntactic model.

camio commented on proposals/p2188.md:
```diff
@@ -0,0 +1,655 @@
+
+### Matching classes with struct patterns
+
+We could allow a class to be matched by a struct patttern that matches its
```
```suggestion
We could allow a class to be matched by a struct pattern that matches its
```

camio reviewed (commented):
It looks like exhaustiveness checking isn't done here. Exhaustiveness checking is a nice feature to have in a language, especially as it intends to supersede C++'s `std::visit` (which does check exhaustiveness). Rust's pattern matching has this feature as well. See [P2211R0](https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2020/p2211r0.pdf) for some possible ways to specify and implement this.

It would be great to see what this syntax looks like with a set of examples. The ones provided in [P1371R3](https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1371r3.pdf) are fairly decent.

Generating bindings to intermediate patterns can sometimes be useful, especially when matching a single field with many-fielded structs.

```c++
// C++
struct Box {
    Point3D corner0;
    Point3D corner1;
};

// Carbon
match( box ) {
  case {.corner1 = {0, _, _}} => { /*how to refer to corner1 here?*/ }
}
```

zygoloid committed:
Fix typo.

Co-authored-by: David Sankel <camior@gmail.com>

zygoloid committed:
Respond to some review comments.

zygoloid committed:
Add as-patterns as future work.

zygoloid commented:
> It looks like exhaustiveness checking isn't done here.

Oops, that's an unintended omission. Added a section on that based on what I implemented in explorer in #2164, which currently doesn't require exhaustiveness but does consider exhaustiveness when seeing if control can reach the end of a value-returning function.

It sounds like you're suggesting that `match` should require exhaustiveness always. If I'm understanding that correctly, it does seem like a simplification of some of the rules, and would probably be better for matches on choice types, at the cost of requiring non-exhaustive matches to be more verbose (eg, adding a `default => {}`), but it's not clear to me what else should be considered in the cost/benefit analysis here. Are there more factors to consider?

> It would be great to see what this syntax looks like with a set of examples. The ones provided in [P1371R3](https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1371r3.pdf) are fairly decent.

Yeah, I agree; while there are examples throughout here, they're all a bit small and a little contrived, and more complete examples like those in P1371 would be great to have. That said, I'm hesitant to copy and translate someone else's work without permission from all the authors. Maybe I can use those as inspiration at least.

> Generating bindings to intermediate patterns can sometimes be useful, especially when matching a single field with many-fielded structs.

I agree; added a future work section on this topic. I think there are enough design questions in that space that we've not thought about yet (mostly syntactic) that I'd prefer to defer that work for a future proposal.

zygoloid committed:
Require exhaustiveness as suggested in code review and discussed on
discord.

See
https://discord.com/channels/655572317891461132/748959784815951963/1037445448908161104

zygoloid commented:
> It sounds like you're suggesting that `match` should require exhaustiveness always. If I'm understanding that correctly, it does seem like a simplification of some of the rules, and would probably be better for matches on choice types, at the cost of requiring non-exhaustive matches to be more verbose (eg, adding a `default => {}`), but it's not clear to me what else should be considered in the cost/benefit analysis here. Are there more factors to consider?

We had some good discussion of this [on discord](https://discord.com/channels/655572317891461132/748959784815951963/1037445448908161104) and concluded that we do want `match` to be exhaustive. That's now part of this proposal.

camio commented:
I looked at all the changes since my prior review and they look good to me.

> I'm hesitant to copy and translate someone else's work without permission from all the authors. Maybe I can use those as inspiration at least.

Most of those examples come from my paper https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0095r1.html. Feel free to use them. The red-black tree example from the newer paper was also written by me, so feel free to use that one as well.

zygoloid committed:
Add examples.

zygoloid commented:
> Most of those examples come from my paper https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0095r1.html. Feel free to use them. The red-black tree example from the newer paper was also written by me, so feel free to use that one as well.

Thanks, added!

zygoloid committed:
Re-add sharper template dependence rule.

josh11b commented on proposals/p2188.md:
```diff
@@ -454,15 +503,179 @@ fn TypeName[template T:! Type](x: T) -> String {
+Some definitions:
+
+-   A pattern _P_ is _useful_ in the context of a set of patterns _C_ if _P_ can
+    match any value that no pattern in _C_ matches.
```
The way I read this definition, it implies that if _P_ is useful, then _C union P_ is exhaustive, which I don't think is intended.

```suggestion
-   A pattern _P_ is _useful_ in the context of a set of patterns _C_ if there exists
    a value that _P_ can match that no pattern in _C_ matches.
```

josh11b commented on proposals/p2188.md:
```diff
@@ -454,15 +503,179 @@ fn TypeName[template T:! Type](x: T) -> String {
+    the context of _C_.
+-   A pattern _P_ is _refutable_ if there are values that it does not match,
+    that is, if the pattern `_` is useful in the context of {_P_}. Equivalently,
+    the pattern _P_ is _refuatble_ if the set of patterns {_P_} is not
```
```suggestion
    the pattern _P_ is _refutable_ if the set of patterns {_P_} is not
```

josh11b commented on proposals/p2188.md:
```diff
@@ -454,15 +503,179 @@ fn TypeName[template T:! Type](x: T) -> String {
+    Equivalently, _C_ is exhaustive if the pattern `_: auto` is not useful in
+    the context of _C_.
+-   A pattern _P_ is _refutable_ if there are values that it does not match,
+    that is, if the pattern `_` is useful in the context of {_P_}. Equivalently,
```
```suggestion
    that is, if the pattern `_: auto` is useful in the context of {_P_}. Equivalently,
```

josh11b commented on proposals/p2188.md:
````diff
@@ -454,15 +503,179 @@ fn TypeName[template T:! Type](x: T) -> String {
+```
+
+```
+fn IsTrue(b: bool) -> bool {
+  match (b) {
+    case false => { return false; }
+    case true => { return true; }
+  }
+  // Code here is considered to be unreachable.
+}
````
Might be helpful to explicitly say that in this example the match is considered exhaustive

zygoloid replied:
```suggestion
fn IsTrue(b: bool) -> bool {
  // Considered exhaustive.
  match (b) {
    case false => { return false; }
    case true => { return true; }
  }
  // Code here is considered to be unreachable.
}
```

josh11b commented on proposals/p2188.md:
```diff
@@ -454,15 +503,179 @@ fn TypeName[template T:! Type](x: T) -> String {
+        with the same type structure are an error unless either an `impl` is
+        provided to cover the overlap or a `match_first` declaration is used to
+        order the `impl`s. (This is a pre-existing rule and is unchanged by this
+        proposal.)
```
I believe the pre-existing rule is stricter: 
```suggestion
    -   For a set of `impl`s that match a given `impl` lookup, argument
        deduction is used rather than pattern matching, but `impl`s with the
        same type structure are an error unless a `match_first` declaration
        is used to order the `impl`s. (This is a pre-existing rule and is
        unchanged by this proposal.)
```

josh11b commented on proposals/p2188.md:
```diff
@@ -454,15 +503,179 @@ fn TypeName[template T:! Type](x: T) -> String {
+
 ## Rationale
 
 -   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
```
It seems like exhaustiveness is partly about forcing needed changes in the presence of evolution, unless the match statement was explicitly written to handle cases that may be added later.

zygoloid replied:
Done.

josh11b commented on proposals/p2188.md:
```diff
@@ -551,12 +812,19 @@ fn I() {
+Having distinct syntax for type-level matching and value-level matching helps
+guide the reader to the correct interpretation, even though the underlying
+matching process is expected to be similar or identical. As a result, we keep
+type deduction syntax and pattern matching syntax separate for now:
```
Address contexts where you can't write `[...]` like `match` statements?

zygoloid replied:
Done.

josh11b commented on proposals/p2188.md:
```diff
@@ -611,6 +879,91 @@ _scrutinee_ `==` _expression_ rather than _expression_ `==` _scrutinee_ -- which
+grammar, and it's not clear that this feature would add sufficient value to
+justify that complexity.
+
+An additional concern is that if we add `let`...`else` syntax, this would
```
Add a reference to this discussion? https://discord.com/channels/655572317891461132/748959784815951963/981691016040030279

zygoloid replied:
Added.

josh11b commented on proposals/p2188.md:
```diff
@@ -611,6 +879,91 @@ _scrutinee_ `==` _expression_ rather than _expression_ `==` _scrutinee_ -- which
+    not computable in general to tell whether a given set of values is
+    exhaustive, so we would only be able to apply this in some subset of cases.
+-   This restriction is straightforward to work around by replacing the final
+    value match with a `default` or `_`.
```
It seems less error-prone to add a `default` than replace a match with a `default`

zygoloid replied:
Added that as a suggested option.

josh11b commented on proposals/p2188.md:
```diff
@@ -624,41 +977,695 @@ match (x) {
+        stream->Print("Fire a laser with {0} intensity.\n", intensity);
+      }
+      case .rotate(degrees: f64) => {
+        stream->Print("Rotate by {9} degrees.\n", degrees);
```
```suggestion
        stream->Print("Rotate by {0} degrees.\n", degrees);
```

zygoloid committed:
Apply suggestions from code review

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

zygoloid committed:
Review comments.

chandlerc commented on proposals/p2188.md:
```diff
@@ -0,0 +1,1689 @@
+extensible, which if used would mean that a set of patterns for that choice type
+would only be considered exhaustive if it includes a wildcard pattern.
+
+## Future work
```
Maybe also mention an expression-form as future work?

zygoloid replied:
Done.

chandlerc commented on proposals/p2188.md:
````diff
@@ -0,0 +1,1689 @@
+}
+```
+
+### Introducer syntax for expression patterns
````
I feel like this section doesn't fully capture what we're giving up without this. Is it possible to capture some of the context from the discussion leading up to this Discord message? https://discord.com/channels/655572317891461132/709488742942900284/1020559500597538886

Generally, I wonder if we want to more deeply revisit whether we can and should shift the syntax in this direction. For example, I think I see plausible solutions to the purely syntactic concerns with identifying an introducer below.

That said, I'd be happy to either move that discussion to an issue for leads, and even do that as a fix-forward with even a minor note here that we anticipate revisiting this specific aspect of the syntax to see if there are approaches that might result in an overall better model.

zygoloid replied:
Added some more notes and links. I favor the fix-forward approach, given that the most promising path involves changing the syntax for `where` expressions to remove the other, conflicting meaning of `is`.

chandlerc commented on proposals/p2188.md:
```diff
@@ -0,0 +1,1689 @@
+[1](https://github.com/carbon-language/carbon-lang/discussions/1495) and
+[2](https://github.com/carbon-language/carbon-lang/discussions/1988).
+
+### Struct pattern syntax
+
+We could omit the `, _` syntax. This would simplify struct patterns, but at the
+cost of removing a feature that can be useful for reducing verbosity and making
+library evolution easier.
```
I'm not fully convinced that we should add this yet TBH.

I wonder if we'll be able to address this usecase *with* variadics, and with tools that also address discarding an unspecified number of tuple elements.

I lean a little bit towards maybe documenting this option but waiting to see if we instead can solve this as part of variadic patterns more generally. What do other folks think here?

The flip side is that I think this is a minor feature and easily *removed* if we find that variadics obviate it, so I also don't think this is a big deal either way.

geoffromer replied:
Note that struct fields aren't expressions or values, so the ability to create and match packs of struct fields would be a separate, bespoke feature, not something that would emerge automatically from a more general variadics design. I don't think we've discussed having such a feature, and it's not clear to me if there are any other use cases for it (maybe metaprogramming?). So my inclination would be to keep `, _` for now, and then replace it if and when we discover we want packs-of-struct-fields for other reasons.

zygoloid replied:
I agree that it's not a big deal either way, because this should be easy to change later. I lean towards adding it and seeing if we can replace it with something else later, such as variadics, and similarly seeing if it gets used heavily.

chandlerc commented on proposals/p2188.md:
```diff
@@ -0,0 +1,1689 @@
+We allow `case`s within a `match` statement to have _guards_. These are not part
+of pattern syntax, but instead are specific to `case` syntax:
+
+-   _case_ ::= `case` _pattern_ [`if` _expression_]? `=>` _block_
```
Idle and pretty minor comment -- I wonder if we should consider alternative syntaxes for the `=>` part of this... Specifically in case we end up wanting `=>` for any other parts of the grammar.

Not super important though, as this might be the *least* impactful syntax question here, and the easiest for us to change our minds about subsequently. =D

zygoloid replied:
This proposal isn't really considering syntax options for `case` / `match` in general; all this section is trying to describe is where the `if` _expression_ part fits into the existing (but not in an approved design) syntax. We should consider alternatives for `match` and `case` syntax, but not in this proposal.

chandlerc commented on proposals/p2188.md:
```diff
@@ -0,0 +1,1689 @@
+with permission from the author of those examples, David Sankel. Thank you,
+David!
+
+### Examples from P0095R1
```
Does it make sense to provide a comparison to the syntax proposed in https://wg21.link/p2392 and more generally acknowledge some of the differences in design there?

zygoloid replied:
If that comparison is useful when evaluating this proposal, I can add it, but I'm a little hesitant to get too deep into comparing this with in-flight WG21 proposals. I'm a little tempted to remove the P0095R1 / P1371R3 comparisons, given that they're not comparing against an existing language.

chandlerc reviewed (commented):
Really awesome to see all of this coming together.

I do still have a few areas I'd like to push on, but to be clear I think all of these are relatively small and narrow aspects of the proposal. Even the syntactic questions, while they'll show up everywhere, are ultimately a small part of the proposal.

zygoloid committed:
Updates based on review comments.

zygoloid committed:
Add another benefit of `is` from the weekly sync meeting.

chandlerc reviewed (approved):
Just officially approving for leads. The future work and other todos cover everything that I think is outstanding, and this seems like a really clear step forward.
